numpy/core/shape_base.py

```diff
@@ -423,32 +423,68 @@ def format_index(index):
         return parent_index, _nx.ndim(arrays)
 
 
-def _block(arrays, max_depth, result_ndim):
+def _assignment_recursion(result, arrays, list_ndim, start_indicies,
+                          depth=0):
+    if depth < list_ndim:
+        concat_depth = depth + result.ndim - list_ndim
+        for arr in arrays:
+            shape = _assignment_recursion(result, arr, list_ndim,
+                                          start_indicies, depth=depth+1)
+            start_indicies[concat_depth] += shape[concat_depth]
+        # lists are mutable, so we have to reset the start_indicies list
+        start_indicies[concat_depth:] = (0, ) * (result.ndim - concat_depth)
+        return shape
+    else:
+        # We've bottom out, assign the array
+        slices = tuple(slice(start, start+size)
+                       for start, size in zip(start_indicies, arrays.shape))
+        result[slices] = arrays
+        return arrays.shape
+
+
+def _shape_recursion(arrays, list_ndim, result_ndim, depth=0):
+    """Internal implementation that gets the shape of the final array.
     """
-    Internal implementation of block. `arrays` is the argument passed to
-    block. `max_depth` is the depth of nested lists within `arrays` and
-    `result_ndim` is the greatest of the dimensions of the arrays in
-    `arrays` and the depth of the lists in `arrays` (see block docstring
-    for details).
+    if depth < list_ndim:
+        shapes = [_shape_recursion(arr, list_ndim, result_ndim, depth+1)
+                  for arr in arrays]
+        concat_depth = depth + result_ndim - list_ndim
+        for shape in shapes[1:]:
+            if (shape[:concat_depth] != shapes[0][:concat_depth] or
+                    shape[concat_depth+1:] != shapes[0][concat_depth+1:]):
+                raise ValueError('Mismatched array shapes in block.')
+        shape_on_dim = sum(shape[concat_depth] for shape in shapes)
+        # Take a shape, any shape
+        shape = shapes[0]
+        shape = shape[:concat_depth] + (shape_on_dim,) + shape[concat_depth+1:]
+        return shape
+    else:
+        # We've 'bottomed out'
+        return arrays.shape
+
+
+def _dtype_recursion(arrays, list_ndim, depth=0):
+    """Internal implementation that gets the dtype of the final array.
     """
-    def atleast_nd(a, ndim):
```

Comment:
I mean this `atleast_nd` function, which already does allow subclasses

Reply:
I'll keep it if you insist. I just really don't like 1 liner, single caller internal functions. We could just make this function public too

Reply:
There's a PR somewhere about making it public. The nice thing about having it internal like this is that if we do decide to make it public, we already know where we can leverage it internally.

Reply:
ok.